ngrinder-controller/src/main/java/org/ngrinder/agent/service/ClusteredAgentManagerService.java

```diff
@@ -82,112 +76,48 @@ public ClusteredAgentManagerService(AgentManager agentManager, AgentManagerRepos
-		for (AgentIdentity agentIdentity : allAttachedAgents) {
-			AgentControllerIdentityImplementation existingAgent = cast(agentIdentity);
-			attachedAgentMap.put(createKey(existingAgent), existingAgent);
-		}
-		Map<String, AgentInfo> agentsInDBMap = Maps.newHashMap();
-		// step1. check all agents in DB, whether they are attached to
-		// controller.
-		for (AgentInfo eachAgentInDB : getAllLocal()) {
-			String keyOfAgentInDB = createKey(eachAgentInDB);
-			agentsInDBMap.put(keyOfAgentInDB, eachAgentInDB);
-			AgentControllerIdentityImplementation agentIdentity = attachedAgentMap.remove(keyOfAgentInDB);
-			if (agentIdentity != null) {
-				// if the agent attached to current controller
-				if (!isCurrentRegion(agentIdentity)) {
-					if (eachAgentInDB.getState() != WRONG_REGION) {
-						eachAgentInDB.setApproved(false);
-						eachAgentInDB.setRegion(getConfig().getRegion());
-						eachAgentInDB.setState(WRONG_REGION);
-						updatedAgents.add(eachAgentInDB);
-					}
-				} else if (!hasSameInfo(eachAgentInDB, agentIdentity)) {
-					fillUp(eachAgentInDB, agentIdentity);
-					updatedAgents.add(eachAgentInDB);
-				} else if (!hasSameState(eachAgentInDB, agentIdentity)) {
-					eachAgentInDB.setState(getAgentManager().getAgentState(agentIdentity));
-					stateUpdatedAgents.add(eachAgentInDB);
-				} else if (eachAgentInDB.getApproved() == null) {
-					updatedAgents.add(fillUpApproval(eachAgentInDB));
-				}
-			} else { // the agent in DB is not attached to current controller
-				if (eachAgentInDB.getState() != INACTIVE) {
-					eachAgentInDB.setState(INACTIVE);
-					stateUpdatedAgents.add(eachAgentInDB);
-				}
-			}
-		}
-
-		// step2. check all attached agents, whether they are new, and not saved
-		// in DB.
-		for (AgentControllerIdentityImplementation agentIdentity : attachedAgentMap.values()) {
-			AgentInfo agentInfo = agentManagerRepository.findByIpAndName(
-					agentIdentity.getIp(),
-					agentIdentity.getName());
-			if (agentInfo == null) {
-				agentInfo = new AgentInfo();
-				newAgents.add(fillUp(agentInfo, agentIdentity));
-			} else {
-				updatedAgents.add(fillUp(agentInfo, agentIdentity));
-			}
-			if (!isCurrentRegion(agentIdentity)) {
-				agentInfo.setState(WRONG_REGION);
-				agentInfo.setApproved(false);
-			}
-		}
-
-		cachedLocalAgentService.updateAgents(newAgents, updatedAgents, stateUpdatedAgents, null);
-		if (!newAgents.isEmpty() || !updatedAgents.isEmpty()) {
-			expireLocalCache();
-		}
+	public List<AgentInfo> getAllVisible() {
+		return getAllLocal()
+			.stream()
+			.filter(this::activeRegionOnly)
+			.collect(toList());
 	}
 
-	public List<AgentInfo> getAllActive() {
-		return filterOnlyActiveRegion(agentManagerRepository.findAll(active()));
+	@Override
+	protected Set<AgentInfo> getVisibleAsSet() {
+		return getAllLocal()
+			.stream()
+			.filter(agentInfo -> agentInfo.getRegion().equals(getConfig().getRegion()))
```

Comment:
Why do we need this filter? Visible is same as all local?

Reply:
In cluster mode allLocal() return All agents information in IMap.
So use filter to get agents in particular region.

need to rename methods